src/JWTVerifier.php

```diff
@@ -5,160 +5,294 @@
-            // validate that this JWT was made for us
-            $audience = $decodedToken->aud;
-            if (! is_array($audience)) {
-                $audience = [$audience];
+        if (empty($body_decoded->aud)) {
+            throw new InvalidTokenException('Token audience not found.');
+        }
+
+        // Validate the token audience.
+        $audience = $body_decoded->aud;
+        if (! is_array($audience)) {
+            $audience = [$audience];
+        };
+        if (! count(array_intersect($audience, $this->valid_audiences))) {
+            throw new InvalidTokenException('This token is not intended for us.');
+        }
+
+        if ('HS256' === $head_decoded->alg) {
+            $secret = $this->client_secret;
+        } else {
+            if (empty($head_decoded->kid)) {
+                throw new CoreException('Token key ID is missing for RS256 token.');
             }
 
-            if (count(array_intersect($audience, $this->valid_audiences)) == 0) {
-                throw new InvalidTokenException('This token is not intended for us.');
+            if (empty($body_decoded->iss) || ! in_array($body_decoded->iss, $this->authorized_iss)) {
+                throw new CoreException('We cannot trust on a token issued by `'.$body_decoded->iss.'`.');
             }
+
+            $jwks_url                   = $body_decoded->iss.$this->jwks_path;
+            $secret[$head_decoded->kid] = $this->JWKFetcher->requestJwkX5c($jwks_url, $head_decoded->kid);
+        }
+
+        try {
+            return $this->decodeToken($jwt, $secret);
         } catch (\Exception $e) {
             throw new CoreException($e->getMessage());
         }
+    }
+
+    /**
+     * Wrapper for JWT::decode().
+     *
+     * @param string       $jwt    JWT to decode.
+     * @param string|array $secret Secret to use.
+     *
+     * @return mixed
+     *
+     * @codeCoverageIgnore
+     */
+    protected function decodeToken($jwt, $secret)
+    {
+        return JWT::decode($jwt, $secret, $this->supported_algs);
+    }
 
-        return $decodedToken;
+    /**
+     * Base64 decode a string.
+     *
+     * @param string $encoded Base64 encoded string.
+     *
+     * @return string
+     *
+     * @codeCoverageIgnore
+     */
+    private function decodeB64($encoded)
+    {
+        return JWT::urlsafeB64Decode($encoded);
+    }
+
+    /**
+     * Base64 and JSON decode a string.
+     *
+     * @param string $segment Base64 encoded JSON string.
+     *
+     * @return object
+     *
+     * @codeCoverageIgnore
+     */
+    private function decodeTokenSegment($segment)
+    {
+        return JWT::jsonDecode(JWT::urlsafeB64Decode($segment));
+    }
+
+    /**
+     * Check whether the $alg parameter is supported.
+     *
+     * @param string $alg Algorithm to check.
+     *
+     * @return boolean
+     *
+     * @codeCoverageIgnore
+     */
+    private function supportsAlg($alg)
```

Comment:
I know you're using this in 2 or 3 places so I get why you moved this to a method. But this sounds more like a "array contains" method. (which you're using in many other places such as when comparing audiences). Consider refactoring this 🆗